consensus/istanbul/backend/engine.go

```diff
@@ -634,10 +634,19 @@ func (sb *backend) snapshot(chain consensus.ChainReader, number uint64, hash com
 		return nil, err
 	}
 	if sb.governance.ProposerPolicy() == uint64(istanbul.WeightedRandom) {
+		pHeader := chain.GetHeaderByNumber(params.CalcStakingBlockNumber(snap.Number + 1))
```

Comment:
Why is this routine added again?

Reply:
The parameter is different. This one is for staking update using `CalcStakingBlockNumber` and the other one is for proposer update using `CalcProposerBlockNumber`.